File: data-structures/assignments/week4/hash_substring/hash_substring.py
# python2


import random


def rabin_karp(text, pattern):
    pattern_hash, hashes = precompute_hashes(pattern, text)
    shifts = []
    for i in xrange(len(text)-len(pattern)+1):
        if hashes[i] == pattern_hash and pattern_matched(pattern, text, i):
            shifts.append(i)
    return shifts


def precompute_hashes(pattern, text):
    p = 1000000007
    x = random.randint(1, p)
    pattern_hash = 0
    for c in reversed(pattern):
        pattern_hash = (ord(c) + x*pattern_hash) % p
    hashes = [0] * len(text)
    y = (x**len(pattern)) % p
    text_hash = 0
    for i in xrange(len(text)-1, len(text)-len(pattern)-1, -1):
        text_hash = (ord(text[i]) + text_hash*x) % p
        hashes[i] = text_hash
    for i in xrange(len(text)-len(pattern)-1, -1, -1):
        hashes[i] = (ord(text[i]) + x*hashes[i+1] - ord(text[i+len(pattern)])*y) % p
    return pattern_hash, hashes


def pattern_matched(pattern, text, i):
    for j, c in enumerate(pattern):
        if c != text[i+j]:
            return False
    return True


def read_data():
    pattern = raw_input()
    text = raw_input()
    return pattern, text


def main():
    pattern, text = read_data()
    shifts = rabin_karp(text, pattern)
    print(' '.join(map(str, shifts)))


if __name__ == '__main__':
    main()

File: data-structures/assignments/week4/hash_substring/hash_substring.cpp
#include <iostream>
#include <vector>


const int p = 1000000007;
const int x = 263;


int polynomial_hash(std::string::const_iterator start, std::string::const_iterator end) {
    unsigned long long hash = 0;
    while (start != end) {
        hash = (*(--end) + x*hash) % p;
    }
    return (int)hash;
}


std::vector<int> precompute_hashes(const std::string &s, int n) {
    int k = (int)s.size()-n+1;
    std::vector<int> hashes(k, 0);
    if (k < 1) return hashes;
    int i = k-1;
    hashes[i] = polynomial_hash(s.begin()+i, s.begin()+i+n);
    long long y = 1;
    for (int j = 0; j < n; j++) {
        y = (y*x) % p;
    }
    while (--i >= 0) {
        hashes[i] = (p + (s[i] + x*(long long)hashes[i+1] - s[i+n]*y) % p) % p;
    }
    return hashes;
}


void show_vector(const std::vector<int> &v) {
    for (int x : v) {
        printf("%d ", x);
    }
    printf("\n");
}


std::vector<int> robin_karp(const std::string &pattern, const std::string &text) {
    std::vector<int> shifts;
    std::vector<int> hashes = precompute_hashes(text, (int)pattern.size());
    int pattern_hash = polynomial_hash(pattern.begin(), pattern.end());
    for (int i = 0; i < text.size()-pattern.size()+1; i++) {
      if ((hashes[i] == pattern_hash) & std::equal(text.begin()+i, text.begin()+i+pattern.size(), pattern.begin())) {
        shifts.push_back(i);
      }
    }
    return shifts;
}


int main() {
    std::string pattern, text;
    std::cin >> pattern >> text;
    show_vector(robin_karp(pattern, text));
    return 0;
}
